File: android/src/main/java/com/getcapacitor/community/audio/AudioAsset.java
package com.getcapacitor.community.audio;

import android.content.res.AssetFileDescriptor;
import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;

import com.getcapacitor.community.audio.queue.QueueController;
import com.getcapacitor.community.audio.queue.QueuePlayer;
import com.getcapacitor.community.audio.queue.QueueTrack;

import java.util.ArrayList;
import java.util.concurrent.Callable;

public class AudioAsset {

    private final String TAG = "AudioAsset";

    private final ArrayList<AudioDispatcher> audioList;
    private int playIndex = 0;
    public final String assetId;
    final NativeAudio owner;
    final QueueController queueController;
    public final QueuePlayer queuePlayer;
    public final QueueTrack queueTrack;
    private boolean onPause = false;

    public AudioAsset(
            NativeAudio owner,
            QueueController queueController,
            QueuePlayer queuePlayer,
            QueueTrack queueTrack,
            AssetFileDescriptor assetFileDescriptor,
            int audioChannelNum,
            float volume
    )
            throws Exception {
        audioList = new ArrayList<>();
        this.owner = owner;
        this.queueController = queueController;
        this.queueTrack = queueTrack;
        this.queuePlayer = queuePlayer;
        this.assetId = queueTrack.getUrl();

        if (audioChannelNum < 0) {
            audioChannelNum = 1;
        }

        for (int x = 0; x < audioChannelNum; x++) {
            AudioDispatcher audioDispatcher = new AudioDispatcher(
                    owner.getActivity(),
                    queueTrack.getUrl(),
                    assetFileDescriptor,
                    volume
            );
            audioList.add(audioDispatcher);
            if (audioChannelNum == 1) audioDispatcher.setOwner(this);
        }
    }

    public void dispatchComplete() {
        this.owner.dispatchComplete(this.assetId);
    }

    public void prepareComplete() {

        this.owner.prepareComplete(this.assetId);

        Log.d(TAG, "prepare complete assetId " + this.assetId + " isPlaying " + this.audioList.get(0).isPlaying());
    }

    public void play(Double time, Callable<Void> callback) throws Exception {
        AudioDispatcher audio = audioList.get(playIndex);

        if (audio != null) {
            audio.play(time, callback);
            playIndex++;
            playIndex = playIndex % audioList.size();
        }
    }

    public double getDuration() {
        Log.d(TAG, "myLog getDuration");
        if (audioList.size() != 1) return 0;

        AudioDispatcher audio = audioList.get(playIndex);

        if (audio != null) {
            return audio.getDuration();
        }
        return 0;
    }

    public double getCurrentPosition() {
        if (audioList.size() != 1) return 0;

        AudioDispatcher audio = audioList.get(playIndex);

        if (audio != null) {
            return audio.getCurrentPosition();
        }
        return 0;
    }

    public boolean pause() throws Exception {
        boolean wasPlaying = false;

        for (int x = 0; x < audioList.size(); x++) {
            AudioDispatcher audio = audioList.get(x);
            wasPlaying |= audio.pause();
        }
        onPause = true;
        return wasPlaying;
    }

    public boolean isPaused() {
        return onPause;
    }

    public void resume() throws Exception {
        if (audioList.size() > 0) {
            AudioDispatcher audio = audioList.get(0);

            if (audio != null) {
                audio.resume();
            }
        }
        onPause = false;
    }

    public void stop() throws Exception {
        for (int x = 0; x < audioList.size(); x++) {
            AudioDispatcher audio = audioList.get(x);

            if (audio != null) {
                audio.stop();
            }
        }
    }

    public void loop() throws Exception {
        AudioDispatcher audio = audioList.get(playIndex);

        if (audio != null) {
            audio.play(0.0, null);
            audio.loop();
            playIndex++;
            playIndex = playIndex % audioList.size();
        }
    }

    public void unload() throws Exception {
        this.stop();

        for (int x = 0; x < audioList.size(); x++) {
            AudioDispatcher audio = audioList.get(x);

            if (audio != null) {
                audio.unload();
            }
        }

        audioList.clear();
    }

    public void setVolume(float volume) throws Exception {
        Log.d(TAG, "set volume " + volume);
        for (int x = 0; x < audioList.size(); x++) {
            AudioDispatcher audio = audioList.get(x);

            if (audio != null) {
                audio.setVolume(volume);
            }
        }
    }

    public void seek(double time) {
        for (int x = 0; x < audioList.size(); x++) {
            AudioDispatcher audio = audioList.get(x);

            if (audio != null) {
                audio.seekTo(time);
            }
        }
    }

    public boolean isPlaying() throws Exception {
        if (audioList.size() != 1) return false;

        return audioList.get(playIndex).isPlaying();
    }

    public void playingChanged(boolean isPlaying) {
        if (isPlaying) {
            queuePlayer.fadeIn(this);
        }

        queuePlayer.playingChanged(isPlaying);
    }

    public AppCompatActivity getActivity() {
        return owner.getBridge().getActivity();
    }
}

File: ios/Plugin/AudioAsset/AudioAsset.swift
//
//  AudioAsset.swift
//  Plugin
//
//  Created by priyank on 2020-05-29.
//  Copyright © 2020 Max Lynch. All rights reserved.
//

import AVFoundation

public class AudioAsset: NSObject, AVAudioPlayerDelegate {

    var channels: NSMutableArray = NSMutableArray()
    var playIndex: Int = 0
    var assetId: String = ""
    var initialVolume: NSNumber = 1.0

    var trailing = false

    let FADE_STEP: Float = 0.05
    let FADE_DELAY: Float = 0.08

    weak var owner: NativeAudio!
    
    let queueTrack: QueueTrack
    weak var queuePlayer: QueuePlayer!
    
    var isPaused: Bool = false


    init(
        owner: NativeAudio,
        queueTrack: QueueTrack,
        queuePlayer: QueuePlayer,
        withAssetId assetId:String,
        withPath path: String!,
        withChannels channels: NSNumber!,
        withVolume volume: NSNumber!,
        useFade: Bool!
    ) {
        
        self.owner = owner
        self.queueTrack = queueTrack
        self.queuePlayer = queuePlayer
        self.assetId = assetId
        self.channels = NSMutableArray.init(capacity: channels as! Int)
        super.init()

        let pathUrl: NSURL! = NSURL(string: path)

        for _ in 0..<channels.intValue {
            let player: Player! = Player(url: pathUrl as URL, useFade: useFade)

            player.avPlayer.volume = volume.floatValue
            self.channels.addObjects(from: [player as Any])
            if channels == 1 {
                NotificationCenter.default.addObserver(forName: .AVPlayerItemDidPlayToEndTime, object: player.avPlayer.currentItem, queue: .main) { [weak self] _ in
                    guard let self = self else { return }
                    queuePlayer.queueController?.completion(assetId: self.assetId)
                    if trailing {
                        return
                    }
                    owner.nowPlayingContoroller.playerStoppedPlaying(track: self.queueTrack)
                }
            }
        }
    }

    func getCurrentTime() -> TimeInterval {
        if channels.count != 1 {
            return 0
        }

        let player: Player = channels.object(at: playIndex) as! Player
        let seconds = CMTimeGetSeconds(player.avPlayer.currentTime())
        return seconds
    }

    func getDuration() -> TimeInterval {
        if channels.count != 1 {
            return 0
        }

        let player: Player = channels.object(at: playIndex) as! Player
        guard let currentTime = player.avPlayer.currentItem else { return 0 }
        let dur = currentTime.asset.duration
        let seconds = CMTimeGetSeconds(dur)
        return seconds
    } 
     func getList() -> NSMutableArray {

            return channels
        }

    func play(time: TimeInterval) {
        let player: Player = channels.object(at: playIndex) as! Player
        player.avPlayer.currentItem?.seek(to: CMTimeMakeWithSeconds(time, preferredTimescale: 60000), completionHandler: nil)
        player.avPlayer.play()
        playIndex = Int(truncating: NSNumber(value: playIndex + 1))
        playIndex = Int(truncating: NSNumber(value: playIndex % channels.count))
        isPaused = false
        owner.nowPlayingContoroller.playerStartedPlaying(track: self.queueTrack)
    }

    func playWithFade(time: TimeInterval) {
        self.play(time: time)
    }

    func pause() {
        let player: Player = channels.object(at: playIndex) as! Player
        player.avPlayer.pause()
        isPaused = true
        owner.nowPlayingContoroller.playerStoppedPlaying(track: self.queueTrack)
    }

    func resume() {
        let player: Player = channels.object(at: playIndex) as! Player

//        let timeOffset = player.deviceCurrentTime + 0.01
        player.avPlayer.play()
        isPaused = false
        owner.nowPlayingContoroller.playerStartedPlaying(track: queueTrack)
    }

    func stop() {
        for i in 0..<channels.count {
            let player: Player! = channels.object(at: i) as? Player
            player.avPlayer.pause()
//            player.avPlayer.replaceCurrentItem(with: nil)
        }
        owner.nowPlayingContoroller.playerStoppedPlaying(track: self.queueTrack)
    }

    func stopWithFade() {
        stop()
    }
    
    func seekTo(time: Double) {
        let player: Player! = channels.object(at: Int(playIndex)) as? Player
        player.avPlayer.seek(to: CMTime(seconds: time, preferredTimescale: CMTimeScale(1000)))
    }

    func loop() {
        let player: Player! = channels.object(at: Int(playIndex)) as? Player
        player.avPlayer.play()
        player.isLoop = true
        playIndex = Int(truncating: NSNumber(value: playIndex + 1))
        playIndex = Int(truncating: NSNumber(value: playIndex % channels.count))
    }

    func unload() {
        self.stop()

//        for i in 0..<channels.count {
//            var player: AVAudioPlayer! = channels.object(at: i) as? AVAudioPlayer
//
//            player = nil
//        }

        channels = NSMutableArray()
    }

    func setVolume(volume: NSNumber!) {
        for i in 0..<channels.count {
            let player: Player! = channels.object(at: i) as? Player
            player.avPlayer.volume = volume.floatValue
        }
    }

    public func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        NSLog("playerDidFinish")
        queuePlayer.queueController?.completion(assetId: self.assetId)
    }

    func playerDecodeError(player: AVAudioPlayer!, error: NSError!) {
        queuePlayer.queueController?.error(assetId: self.assetId)
    }

    func isPlaying() -> Bool {
        if channels.count != 1 {
            return false
        }

        let player: Player = channels.object(at: playIndex) as! Player

        return player.avPlayer.rate != 0 && player.avPlayer.error == nil
    }

    @objc func notifyStop() {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            NSLog("notify stop")
            let player: Player! = self.channels.object(at: Int(self.playIndex)) as? Player
            if player.isLoop {
                player.avPlayer.play()
            } else {
                self.owner.notifyListeners("complete", data: ["assetId": self.assetId])
            }

        }
    }
}



class Player {
    var avPlayer: AVPlayer!
    var asset: AVAsset!
    var isLoop: Bool = false

    init(url: URL, useFade: Bool) {
        asset = AVAsset(url: url)
        let item = AVPlayerItem(asset: asset)
        if useFade {
            let duration = asset.duration
            let durationInSeconds = CMTimeGetSeconds(duration)
            let params = AVMutableAudioMixInputParameters(track: asset.tracks.first! as AVAssetTrack)
            let firstSecond = CMTimeRangeMake(start: CMTimeMakeWithSeconds(0, preferredTimescale: 1000), duration: CMTimeMakeWithSeconds(1.0, preferredTimescale: 1000))
            let lastSecond = CMTimeRangeMake(start: CMTimeMakeWithSeconds(durationInSeconds - 1.3, preferredTimescale: 1000), duration: CMTimeMakeWithSeconds(1, preferredTimescale: 1000))
            params.setVolumeRamp(fromStartVolume: 0, toEndVolume: 1, timeRange: firstSecond)
            params.setVolumeRamp(fromStartVolume: 1, toEndVolume: 0, timeRange: lastSecond)
            let mix = AVMutableAudioMix()
            mix.inputParameters = [params]
            item.audioMix = mix
        }
        avPlayer = AVPlayer(playerItem: item)
    }
}
